### PR TITLE
CLI: Add `pki verify` command for better TLS certificate troubleshooting

### PR DESCRIPTION
**Pull request by dnsmichi**
This originates from testing the "is CA" functionality in #7835 and provides
a built-in replacement for the OpenSSL binary for printing and verifying Icinga
TLS certificates.

The CLI command also provides enhanced certificate information printing
with version, serial, signature algorithm and subject alt names (SANs).
The same applies to other functionality using this function (save-cert, etc.).

Supported modes:

- Print the TLS certificate
- Print and verify the CA certificate (checks whether this is a CA certificate)
- Verify the TLS certificate if it is signed by the provided CA certificate
- Verify the TLS certificate's common name (CN) matches against the given CN string

Bonus points:

- The exit codes follow the plugin API, so this can be used for automated checks
- On Windows, no-one ever installs the OpenSSL binary to do verification tests.
With providing this from the icinga2.exe binary, troubleshooting becomes even more easy.

### CODE REVIEW

**Comment by dnsmichi** [comment]
Tests from the docs:

#### Print TLS Certificate

Pass the certificate file to the `--cert` CLI command parameter to print its details.
This prints a shorter version of `openssl x509 -in <file> -text`.

```
$ icinga2 pki verify --cert icinga2-agent2.localdomain.crt

information/cli: Printing certificate 'icinga2-agent2.localdomain.crt'

 Version:             3
 Subject:             CN = icinga2-agent2.localdomain
 Issuer:              CN = Icinga CA
 Valid From:          Feb 14 11:29:36 2020 GMT
 Valid Until:         Feb 10 11:29:36 2035 GMT
 Serial:              12:fe:a6:22:f5:e3:db:a2:95:8e:92:b2:af:1a:e3:01:44:c4:70:e0

 Signature Algorithm: sha256WithRSAEncryption
 Subject Alt Names:   icinga2-agent2.localdomain
 Fingerprint:         40 98 A0 77 58 4F CA D1 05 AC 18 53 D7 52 8D D7 9C 7F 5A 23 B4 AF 63 A4 92 9D DC FF 89 EF F1 4C
```

You can also print the `ca.crt` certificate without any further checks using the `--cert` parameter.

#### Print and Verify CA Certificate 

The `--cacert` CLI parameter allows to check whether the given certificate file is a public CA certificate.

```
$ icinga2 pki verify --cacert ca.crt

information/cli: Checking whether certificate 'ca.crt' is a valid CA certificate.

 Version:             3
 Subject:             CN = Icinga CA
 Issuer:              CN = Icinga CA
 Valid From:          Jul 31 12:26:08 2019 GMT
 Valid Until:         Jul 27 12:26:08 2034 GMT
 Serial:              89:fe:d6:12:66:25:3a:c5:07:c1:eb:d4:e6:f2:df:ca:13:6e:dc:e7

 Signature Algorithm: sha256WithRSAEncryption
 Subject Alt Names:
 Fingerprint:         9A 11 29 A8 A3 89 F8 56 30 1A E4 0A B2 6B 28 46 07 F0 14 17 BD 19 A4 FC BD 41 40 B5 1A 8F BF 20

information/cli: OK: CA certificate file 'ca.crt' was verified successfully.
```

In case you pass a wrong certificate, an error is shown and the exit code is `2` (Critical).

```
$ icinga2 pki verify --cacert icinga2-agent2.localdomain.crt

information/cli: Checking whether certificate 'icinga2-agent2.localdomain.crt' is a valid CA certificate.

 Version:             3
 Subject:             CN = icinga2-agent2.localdomain
 Issuer:              CN = Icinga CA
 Valid From:          Feb 14 11:29:36 2020 GMT
 Valid Until:         Feb 10 11:29:36 2035 GMT
 Serial:              12:fe:a6:22:f5:e3:db:a2:95:8e:92:b2:af:1a:e3:01:44:c4:70:e0

 Signature Algorithm: sha256WithRSAEncryption
 Subject Alt Names:   icinga2-agent2.localdomain
 Fingerprint:         40 98 A0 77 58 4F CA D1 05 AC 18 53 D7 52 8D D7 9C 7F 5A 23 B4 AF 63 A4 92 9D DC FF 89 EF F1 4C

critical/cli: CRITICAL: The file 'icinga2-agent2.localdomain.crt' does not seem to be a CA certificate file.
```

#### Verify Certificate is signed by CA Certificate 

Pass the certificate file to the `--cert` CLI parameter, and the `ca.crt` file to the `--cacert` parameter.
Common troubleshooting scenarios involve self-signed certificates and untrusted agents resulting in disconnects.

```
$ icinga2 pki verify --cert icinga2-agent2.localdomain.crt --cacert ca.crt

information/cli: Verifying certificate 'icinga2-agent2.localdomain.crt'

 Version:             3
 Subject:             CN = icinga2-agent2.localdomain
 Issuer:              CN = Icinga CA
 Valid From:          Feb 14 11:29:36 2020 GMT
 Valid Until:         Feb 10 11:29:36 2035 GMT
 Serial:              12:fe:a6:22:f5:e3:db:a2:95:8e:92:b2:af:1a:e3:01:44:c4:70:e0

 Signature Algorithm: sha256WithRSAEncryption
 Subject Alt Names:   icinga2-agent2.localdomain
 Fingerprint:         40 98 A0 77 58 4F CA D1 05 AC 18 53 D7 52 8D D7 9C 7F 5A 23 B4 AF 63 A4 92 9D DC FF 89 EF F1 4C

information/cli:  with CA certificate 'ca.crt'.

 Version:             3
 Subject:             CN = Icinga CA
 Issuer:              CN = Icinga CA
 Valid From:          Jul 31 12:26:08 2019 GMT
 Valid Until:         Jul 27 12:26:08 2034 GMT
 Serial:              89:fe:d6:12:66:25:3a:c5:07:c1:eb:d4:e6:f2:df:ca:13:6e:dc:e7

 Signature Algorithm: sha256WithRSAEncryption
 Subject Alt Names:
 Fingerprint:         9A 11 29 A8 A3 89 F8 56 30 1A E4 0A B2 6B 28 46 07 F0 14 17 BD 19 A4 FC BD 41 40 B5 1A 8F BF 20

information/cli: OK: Certificate with CN 'icinga2-agent2.localdomain' is signed by CA.
```

#### Verify Certificate matches Common Name (CN)

This allows to verify the common name inside the certificate with a given string parameter.
Typical troubleshooting involve upper/lower case CNs (Windows).

```
$ icinga2 pki verify --cert icinga2-agent2.localdomain.crt --cn icinga2-agent2.localdomain

information/cli: Verifying common name (CN) 'icinga2-agent2.localdomain in certificate 'icinga2-agent2.localdomain.crt'.

 Version:             3
 Subject:             CN = icinga2-agent2.localdomain
 Issuer:              CN = Icinga CA
 Valid From:          Feb 14 11:29:36 2020 GMT
 Valid Until:         Feb 10 11:29:36 2035 GMT
 Serial:              12:fe:a6:22:f5:e3:db:a2:95:8e:92:b2:af:1a:e3:01:44:c4:70:e0

 Signature Algorithm: sha256WithRSAEncryption
 Subject Alt Names:   icinga2-agent2.localdomain
 Fingerprint:         40 98 A0 77 58 4F CA D1 05 AC 18 53 D7 52 8D D7 9C 7F 5A 23 B4 AF 63 A4 92 9D DC FF 89 EF F1 4C

information/cli: OK: CN 'icinga2-agent2.localdomain' matches certificate CN 'icinga2-agent2.localdomain'.
```

In the example below, the certificate uses an upper case CN.

```
$ icinga2 pki verify --cert icinga2-agent2.localdomain.crt --cn icinga2-agent2.localdomain

information/cli: Verifying common name (CN) 'icinga2-agent2.localdomain in certificate 'icinga2-agent2.localdomain.crt'.

 Version:             3
 Subject:             CN = ICINGA2-agent2.localdomain
 Issuer:              CN = Icinga CA
 Valid From:          Feb 14 11:29:36 2020 GMT
 Valid Until:         Feb 10 11:29:36 2035 GMT
 Serial:              12:fe:a6:22:f5:e3:db:a2:95:8e:92:b2:af:1a:e3:01:44:c4:70:e0

 Signature Algorithm: sha256WithRSAEncryption
 Subject Alt Names:   ICINGA2-agent2.localdomain
 Fingerprint:         40 98 A0 77 58 4F CA D1 05 AC 18 53 D7 52 8D D7 9C 7F 5A 23 B4 AF 63 A4 92 9D DC FF 89 EF F1 4C

critical/cli: CRITICAL: CN 'icinga2-agent2.localdomain' does NOT match certificate CN 'icinga2-agent2.localdomain'.
```

**Comment by dnsmichi** [comment]
cc @widhalmt

**Comment by mcktr** [comment]
# Review

Wow thanks this is really helpful! This  worked as expected.  :+1:  I found a few edge cases which brought some minor issues.

## Code

### Deprecated function

```
/home/michael/Development/Icinga/icinga2/lib/base/tlsutility.cpp: In function ‘icinga::Array::Ptr icinga::GetSubjectAltNames(const std::shared_ptr<x509_st>&)’:
/home/michael/Development/Icinga/icinga2/lib/base/tlsutility.cpp:875:60: warning: ‘unsigned char* ASN1_STRING_data(ASN1_STRING*)’ is deprecated [-Wdeprecated-declarations]
    String san = Convert::ToString(ASN1_STRING_data(asn1_str));
                                                            ^
In file included from /usr/include/openssl/e_os2.h:13,
                 from /usr/include/openssl/ssl.h:15,
                 from /home/michael/Development/Icinga/icinga2/lib/base/tlsutility.hpp:11,
                 from /home/michael/Development/Icinga/icinga2/lib/base/tlsutility.cpp:3:
/usr/include/openssl/asn1.h:554:1: note: declared here
 DEPRECATEDIN_1_1_0(unsigned char *ASN1_STRING_data(ASN1_STRING *x))
 ^~~~~~~~~~~~~~~~~~
/home/michael/Development/Icinga/icinga2/lib/base/tlsutility.cpp:875:60: warning: ‘unsigned char* ASN1_STRING_data(ASN1_STRING*)’ is deprecated [-Wdeprecated-declarations]
    String san = Convert::ToString(ASN1_STRING_data(asn1_str));
                                                            ^
In file included from /usr/include/openssl/e_os2.h:13,
                 from /usr/include/openssl/ssl.h:15,
                 from /home/michael/Development/Icinga/icinga2/lib/base/tlsutility.hpp:11,
                 from /home/michael/Development/Icinga/icinga2/lib/base/tlsutility.cpp:3:
/usr/include/openssl/asn1.h:554:1: note: declared here
 DEPRECATEDIN_1_1_0(unsigned char *ASN1_STRING_data(ASN1_STRING *x))
 ^~~~~~~~~~~~~~~~~~
```

The function `ASN1_STRING_data()` is deprecated, we should use `ASN1_STRING_get0_data()` instead. Unfortunately `ASN1_STRING_get0_data()` is not available on Debian jessie, may we use if/else here to detect the used OpenSSL version? 

### Uncaught Exceptions

```
# icinga2 pki verify --cert cat.cert

critical/SSL: Error on bio X509 AUX reading pem file 'cat.cert': 33558530, "error:02001002:system library:fopen:No such file or directory"
critical/Application: Error: std::exception


Additional information is available in '/usr/local/icinga2/var/log/icinga2/crash/report.1582046739.759753'

Abgebrochen
```

```
# icinga2 pki verify --cacert dog.crt

critical/SSL: Error on bio X509 AUX reading pem file 'dog.crt': 33558530, "error:02001002:system library:fopen:No such file or directory"
critical/Application: Error: std::exception


Additional information is available in '/usr/local/icinga2/var/log/icinga2/crash/report.1582046773.322028'

Abgebrochen
```

The function `GetX509Certificate()` should be in a try/catch block.

## Functionality

### Running without parameter

```
# icinga2 pki verify
```

Running `icinga2 pki verify` without any parameters returns nothing. Can we change the behavior so that a warning will be displayed e.g. `Please specifiy one of the following command options: --cert or --cacert`

A similiar behavior occurs if one only sets the `--cn` parameter.

```
# icinga2 pki verify --cn catdog
```

No output.

###  Idea: Passing CA certificate

```
# icinga2 pki verify --cert var/lib/icinga2/ca/ca.crt
information/cli: Printing certificate 'var/lib/icinga2/ca/ca.crt'

 Version:             3
 Subject:             CN = Icinga CA
 Issuer:              CN = Icinga CA
 Valid From:          Feb 18 17:16:25 2020 GMT
 Valid Until:         Feb 14 17:16:25 2035 GMT
 Serial:              6c:7c:b1:1f:33:9c:46:26:1a:b1:69:9f:9b:dd:96:1a:09:21:7d:60

 Signature Algorithm: sha256WithRSAEncryption
 Subject Alt Names:   
 Fingerprint:         3A EC 29 61 0E 15 9A 35 0F C7 CB 22 A1 BA DA 7B 2A 0C 5E 70 9A 93 40 25 6C D0 6B 56 93 EB CF EB 

```

You can pass a CA certificate to the `--cert` parameter. While this is okay I thought about the function `IsCa()` which was recently implemented by you. Can we have a log message here which says `this is a CA certificate you might wanted to use --cacert`? I am unsure if this would be too verbose.

File: lib/cli/pkiverifycommand.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+	if (vm.count("cert"))
+		certFile = vm["cert"].as<std::string>();
+
+	if (vm.count("cacert"))
+		caCertFile = vm["cacert"].as<std::string>();
+
+	/* Verify CN in certificate. */
+	if (!cn.IsEmpty() && !certFile.IsEmpty()) {
+		std::shared_ptr<X509> cert = GetX509Certificate(certFile);
```
Comment (mcktr) [inline lib/cli/pkiverifycommand.cpp]: Uncaught exception.

File: lib/cli/pkiverifycommand.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+			Log(LogCritical, "cli")
+				<< "CRITICAL: CN '" << cn << "' does NOT match certificate CN '" << certCN << "'.";
+
+			return ServiceCritical;
+		}
+	}
+
+	/* Verify certificate. */
+	if (!certFile.IsEmpty() && !caCertFile.IsEmpty()) {
+		std::shared_ptr<X509> cert = GetX509Certificate(certFile);
```
Comment (mcktr) [inline lib/cli/pkiverifycommand.cpp]: Uncaught exception.

File: lib/cli/pkiverifycommand.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		std::shared_ptr<X509> cacert = GetX509Certificate(caCertFile);
```
Comment (mcktr) [inline lib/cli/pkiverifycommand.cpp]: Uncaught exception.

File: lib/remote/jsonrpcconnection-pki.cpp [hunk 4]
```diff
@@ -53,7 +53,13 @@ Value RequestCertificateHandler(const MessageOrigin::Ptr& origin, const Dictiona
...
 	String cn = GetCertificateCN(cert);
 
-	bool signedByCA = VerifyCertificate(cacert, cert);
+	bool signedByCA;
+
+	try {
+		signedByCA = VerifyCertificate(cacert, cert);
+	} catch (const std::exception& ex) {
+
+	}
```
Comment (mcktr) [inline lib/remote/jsonrpcconnection-pki.cpp]: Is the `catch` block empty on purpose? If so can you add a short code comment about the swallowed exception?

**Comment by dnsmichi** [comment]
Thanks for the detailed review.

- Exceptions, will be fixed
- Error handling for cn+cert - will be fixed.
- Nothing provided - logging the --help indicator is the least invasive change here.
- Deprecation - I love OpenSSL, but yeah, I'll look into this.

**Comment by dnsmichi** [comment]
@mcktr Everything has been fixed. In terms of passing a CA certificate to `--cert`, this is legit to print its  details. If you really want to verify `isCA`, users need to use `--cacert` then. I would like to avoid doing the check inside `--cert` as this replaces functionality.

**Comment by mcktr** [comment]
@dnsmichi Thanks! I re-tested this an everything is fine :+1:.

> In terms of passing a CA certificate to `--cert`, this is legit to print its details.

I agree with you here.

**Comment by dnsmichi** [comment]
@mcktr Thanks for the final review, I'll add this to 2.12 as troubleshooting help.

**Pull request merged**